### PR TITLE
Feature/decompose compile task

### PR DESCRIPTION
**Pull request by Hrom131**
- [ ] Since this PR suggests a **bug fix**, the relevant tests have been added.
- [x] Since this PR introduces a **new feature**, the update has been discussed in an Issue or with the team.
- [ ] This PR is just a **minor change**, like a typo fix.

---

### CODE REVIEW

File: src/compile/core/CompilationProcessor.ts [hunk 1]
```diff
@@ -18,46 +21,36 @@ import {
...
-import { PtauDownloader } from "../utils/PtauDownloader";
-import { getNormalizedFullPath, renameFilesRecursively, readDirRecursively } from "../../utils/path-utils";
-import { HardhatZKitError } from "../../errors";
-import { NODE_MODULES, PTAU_FILE_REG_EXP } from "../../constants";
-import { Reporter } from "../../reporter";
+import { ZKitConfig } from "../../types/zkit-config";
+import { ArtifactsFileType, CircuitArtifact, ICircuitArtifacts } from "../../types/circuit-artifacts";
 
-export class CompilationProcessor {
+export class CompilationProcessorNew {
```
Comment (Arvolear) [inline src/compile/core/CompilationProcessor.ts:27]:
```suggestion
export class CompilationProcessor {
```

?

File: src/index.ts [hunk 2]
```diff
@@ -12,51 +13,80 @@ import { CircuitTypesGenerator } from "@solarity/zktype";
...
 
 extendEnvironment((hre) => {
   hre.zkit = lazyObject(() => {
+    const circuitArtifacts: CircuitArtifacts = new CircuitArtifacts(
+      getNormalizedFullPath(hre.config.paths.root, hre.config.zkit.compilationSettings.artifactsDir),
+    );
+
     return {
+      circuitArtifacts,
+
```
Comment (Arvolear) [inline src/index.ts:63]:
```suggestion
```

File: src/utils/path-utils.ts [hunk 3]
```diff
@@ -92,3 +93,45 @@ function readdirSync(absolutePathToDir: string) {
...
+//     const parentTokens = parent.split(path.posix.sep).filter((i) => i.length);
+//     const childTokens = child.split(path.posix.sep).filter((i) => i.length);
+
+//     return parentTokens.every((t, i) => childTokens[i] === t);
+//   };
+
+//   return pathList.some((p: any) => {
+//     return isSubPath(getNormalizedFullPath(circuitsRoot, p), source);
+//   });
+// };
```
Comment (Arvolear) [inline src/utils/path-utils.ts:137]:
```suggestion
```

File: src/task-names.ts [hunk 4]
```diff
@@ -1,3 +1,5 @@
 export const TASK_CIRCUITS_COMPILE = "zkit:compile";
+export const TASK_CIRCUITS_COMPILE_SHALLOW = "zkit:compile-shallow";
+export const TASK_CIRCUITS_SETUP = "zkit:setup";
```
Comment (Arvolear) [inline src/task-names.ts:3]: What do you think about this renaming? To make everything one-word.

```suggestion
export const TASK_CIRCUITS_MAKE = "zkit:make";
export const TASK_CIRCUITS_COMPILE = "zkit:compile";
export const TASK_CIRCUITS_SETUP = "zkit:setup";
```

File: src/setup/SetupFilesResolver.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      if (
+        !fsExtra.existsSync(this._circuitArtifacts.getCircuitArtifactFileFullPath(setupInfo.circuitArtifact, "zkey")) ||
+        !fsExtra.existsSync(this._circuitArtifacts.getCircuitArtifactFileFullPath(setupInfo.circuitArtifact, "vkey"))
+      ) {
+        CircuitsSetupCache!.removeEntry(setupInfo.circuitArtifactFullPath);
+      }
+    }
+  }
+
+  protected _needsCompilation(circuitSetupInfo: CircuitSetupInfo, contributionSettings: ContributionSettings): boolean {
```
Comment (Arvolear) [inline src/setup/SetupFilesResolver.ts:97]:
```suggestion
  protected _needsSetup(circuitSetupInfo: CircuitSetupInfo, contributionSettings: ContributionSettings): boolean {
```

**Comment by Arvolear** [comment]
Let's rename `compile` dir to `core` and `core` dir to `compile`. Then move `setup` to `core`.

File: src/task-names.ts (hunk 4 above)
Comment (Hrom131) [reply src/task-names.ts:3]: Sounds good

**Review by Arvolear — APPROVED** [review]
LGTM!

**Pull request merged**